SU-HSDO/suhumsci
# STN-628:  Caption/credit text styles limit

# READY FOR REVIEW

## Summary
Limit the Caption/Credit Text Styles that print to prevent Site Owners from breaking the caption/credits view mode display styles with additional styling. Also adjusting for styles and elements that may be brought into the credits section from Drupal 7 that has unwanted styles or elements.

**Implementation Details:**
We limited the caption/credits field to accept Minimal HTML, which includes the following styles `<em> <strong> <cite> <blockquote cite> <code> <ul type> <ol start type> <li> <dl> <dt> <dd> <h2 id> <h3 id> <h4 id> <h5 id> <h6 id> <hr> <p class=""> <span class=""> <table class=""> <br> <a href hreflang data-entity-substitution data-entity-type data-entity-uuid title class=" hs-button hs-button--big hs-secondary-button hs-more-link hs-external-link hs-file-link hs-pdf-link hs-private-link">`

We would like to allow the following styles (for all other styles we do not want to print the styles, but print the content) for the view mode:
```
<em>
<strong>
<p>
<br>
<a>
```

**To accomplish this we need to:**
Target all text styles not in the above list and use the mixin `hb-text-reset`
Target all non text styles in the above list and set `display: none;`


My approach:
I wanted to utilize the `hb-text-reset` for all the unwanted elements and styles and then add in what we would like to allow. Those items include not just the elements we want to keep but allowing elements to keep their base style identity and mix in with the credits default look. The `<a>` tags to still be identified as links while still keeping the accessibility needed within the dark gradient overlay of the credits area. The strong tag, was reset but added a bold weight back to it to keep it's emphasized identity. Some elements were easy to remove such as `<hr>`, but most others needed to be reset and the content retained, which required some heavy-handedness in a couple of areas as well, so we did have to add a couple of`!important` attributes for those items.

Additionally, since we were adding padding originally to the p tag within the caption container, the overall look was reliant upon that. I reworked this now so the container is taking care of the padding and the child elements are able to display inline next to one another.  Additionally, since the child elements float: right, I added a `flex-end` with `display: flex` to the container so that the elements will display in the order they were added to the WYSIWYG.

Jira Card: [#628](https://sparkbox.atlassian.net/browse/STN-628)

## Need Review By (Date)
01/25

## Urgency
'Medium'

## Steps to Test
1. `npm run test`
2. To test this PR you will first need to turn on the caption/credits to display in the View Mode. You can find this [here](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/display), making sure the "Caption/Credit" field is enabled.  Additionally you may need to make sure it enabled in the [Form Display](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/form-display) if is not already.
3. You will also want to turn on FULL HTML for the credit/caption WYSIWYG even though this option will not be enabled for users later, it will allow you to enter these elements for testing using the style buttons as well as Source access. [You can add the FULL HTML option to credit/captions here.](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/fields/media.image.field_media_image_caption)
4. Once those items are turned on you can start adding credits/captions to your media images you will be testing. The easiest method for this is to go to the [Media Grid page](http://sparkbox-sandbox.suhumsci.loc/admin/content/media-grid) and choose any images you are going to use in testing, switch your WYSIWYG entry to full and add in your elements on the credit/caption field. Doing this here will allow you to choose your images you've edited when adding postcards. I tested two or three images in postcards at a time with different elements.
5. Now, all you have left to do is to[ create Flexible Page](http://sparkbox-sandbox.suhumsci.loc/node/add/hs_basic_page) and add as many postcards to the page with your media images you've added caption/credits to for testing.
I've included some html elements that I used for testing below for easy copy and pasting into Source input, if you'd like to use it:

```
<h1>This is an H1</h1>
<h2>This is an H2</h2>
<h3>This is an H3</h3>
<h4 id="">This is an H4</h4>
<h5 id="">This is an H5</h5>
<h6 id="">This is an H6</h6>
<p>This is a standard caption p tag.</p>
<p><span>This is a span</span></p>
<p><a href="http://google.com">This is an anchor</a></p>
<p><strong>This is strong</strong></p>
<p><a class="hs-pdf-link" href="&lt;front&gt;">This is a pdf link</a></p>
<p><a class="hs-file-link" href="&lt;front&gt;">This is a file link</a></p>
<p><a class="hs-private-link" href="&lt;front&gt;">This is a private link</a></p>
<p><a class="hs-more-link" href="&lt;front&gt;">This is a more link</a></p>
<p><a class="hs-secondary-button" href="&lt;front&gt;">This is a secondary button</a></p>
<p><a class="hs-button--big" href="&lt;front&gt;">This is a big primary button</a></p>
<p><a class="hs-button" href="&lt;front&gt;">This is a primary button</a></p></p>
<strong>This is another strong tag</strong>
<hr />
<p><cite>This is cite</cite></p>
<blockquote cite="">This is a blockquote</blockquote>
<p><code>This is code</code></p>
<ul>
	<li>This is an li</li>
</ul>
<dl>
	<dd>This is a dd</dd>
	<dt>This is a dt</dt>
</dl>
<p><em>This is an em.</em></p>
```

_Note: Some of these elements are within `<p>` tags, which the WYSIWYG editor will add by default upon saving, feel free to remove them from the `<p>` containers and test them individually as well. You can do this by saving your page while still viewing in SOURCE view, it saves before the CKEDITOR has a chance to tamper with your code. If you have any questions on this please let me know._

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)
